File: otters/batcher.py
import time
import pyarrow as pa

from .schema import Schema


class Batcher:
    _schema: pa.Schema
    _buffer: list[dict]
    _batch_size: int
    _flush_interval: float
    _last_flush: float

    def __init__(self, schema: Schema, batch_size: int = 500, flush_ms: int = 50) -> None:
        if batch_size < 1:
            raise ValueError(f"batch_size must be >= 1, got {batch_size}")
        if flush_ms < 1:
            raise ValueError(f"flush_ms must be >= 1, got {flush_ms}")
        
        self._schema = schema.to_arrow()
        self._buffer = []
        self._batch_size = batch_size
        self._flush_interval = flush_ms / 1000.0
        self._last_flush = time.monotonic()
    
    def push(self, row: dict) -> pa.RecordBatch | None:
        self._buffer.append(row)

        if len(self._buffer) >= self._batch_size:
            return self._flush()

        if time.monotonic() - self._last_flush >= self._flush_interval:
            return self._flush()

        return None

    def flush_remaining(self) -> pa.RecordBatch | None:
        return self._flush()
    
    def _flush(self) -> pa.RecordBatch | None:
        if not self._buffer:
            return None

        batch = pa.RecordBatch.from_pylist(self._buffer, schema=self._schema)
        self._buffer.clear()
        self._last_flush = time.monotonic()
        return batch

File: src/batcher.rs
use pyo3::prelude::*;
use pyo3::types::PyList;
use crossbeam_channel::{Receiver, Sender};
use arrow::record_batch::RecordBatch;
use arrow::pyarrow::FromPyArrow;

pub struct Batcher {
    batch_size: usize,
    schema_fields: Vec<String>,
}

impl Batcher {
    pub fn new(batch_size: usize, schema_fields: Vec<String>) -> Self {
        Self { batch_size, schema_fields }
    }
}

/// spawn batcher thread
/// 
/// receives pytho dicts from source thread,
/// buffers them, and then flushes as Arrow RecordBatches
/// 
/// channel still carries PY<PyAny> on both sides
pub fn spawn_batcher(
    receiver: Receiver<Py<PyAny>>,
    sender: Sender<RecordBatch>,
    batch_size: usize,
) -> std::thread::JoinHandle<()> {

    // straight forward buffer batching stuff
    std::thread::spawn(move || {
        let mut buffer: Vec<Py<PyAny>> = Vec::with_capacity(batch_size);

        loop {
            match receiver.recv() {
                Ok(item) => buffer.push(item),
                Err(_) => {
                    if !buffer.is_empty() {
                        if let Some(batch) = flush(&buffer) {
                            sender.send(batch).ok();
                        }
                    }
                    break;
                }
            }

            while buffer.len() < batch_size {
                match receiver.try_recv() {
                    Ok(item) => buffer.push(item),
                    Err(_) => break,
                }
            }

            if buffer.len() >= batch_size {
                if let Some(batch) = flush(&buffer) {
                    sender.send(batch).ok();
                    buffer.clear();
                }
            }
        }
    })
}

///converts buffer of py dicts to single arrow recordbatch
/// aquires gil once per buffer/batch
fn flush(rows: &[Py<PyAny>]) -> Option<RecordBatch> {
    Python::attach(|py| {
        let lst = PyList::new(py, rows.iter().map(|r| r.bind(py))).ok()?;
        let pa = py.import("pyarrow").ok()?;
        let rb_class = pa.getattr("RecordBatch").ok()?;
        let py_batch = rb_class.call_method1("from_pylist", (lst,)).ok()?;
        RecordBatch::from_pyarrow_bound(&py_batch).ok()
    })
}